src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1292,11 +1292,21 @@ String IMergeTreeDataPart::getRelativePathForPrefix(const String & prefix) const
         * This is done only in the case of `to_detached`, because it is assumed that in this case the exact name does not matter.
         * No more than 10 attempts are made so that there are not too many junk directories left.
         */
+
+    auto full_relative_path = fs::path(storage.relative_data_path);
+    if (detached)
+        full_relative_path /= "detached";
+    /// TODO What shall we do with projections?
```

Comment:
It's not possible to detach projection parts alone. I think we can add a logical exception check here.